# Running Table Tests

`table/table.go`

```go
package table

import (
	"errors"
	"fmt"
)

func DoMath(num1, num2 int, op string) (int, error) {
	switch op {
	case "+":
		return num1 + num2, nil
	case "-":
		return num1 - num2, nil
	case "*":
		return num1 + num2, nil
	case "/":
		if num2 == 0 {
			return 0, errors.New("division by zero")
		}
		return num1 / num2, nil 
	default:
		return 0, fmt.Errorf("unknown operator %s", op)
	}
}
```

`table/table_test.go`

- Instead of checking for different branches, we can avoid the repetition by declaring a slice of anonymous structs.
- Next we invoke the `Run` method which takes two params:
  - a name for the subtest 
  - a function with a single parameter of type `*testing.T`.

```go
package table

import "testing"

func TestDoMathTable(t *testing.T){
	data := []struct{
		name string
		num1 int 
		num2 int 
		op string 
		expected int 
		errMsg string 
	}{
		{"addition", 2, 2, "+", 4, ""},
		{"subtraction", 2, 2, "-", 0, ""},
		{"multiplication", 2, 3, "*", 6, ""},
		{"division", 2, 2, "/", 1, ""},
		{"bad_division", 2, 0, "/", 0, `division by zero`},
	}

    // If the error has a custom type or a named sentinel error, use errors.Is or errors.As 
    // to check that the correct error is returned
	for _, d := range data{
		t.Run(d.name, func(t *testing.T){
			result, err := DoMath(d.num1, d.num2, d.op)
			if result != d.expected{
				t.Errorf("Expected %d, got %d", d.expected, result)
			}
			var errMsg string 
			if err != nil{
				errMsg = err.Error() 
			}
			if errMsg != d.errMsg{
				t.Errorf("Expected error message `%s`, got `%s`", d.errMsg, errMsg)
			}
		})
	}
}
```

`Output (failed test)`

```sh
table ➤ go test
--- FAIL: TestDoMathTable (0.00s)
    --- FAIL: TestDoMathTable/multiplication (0.00s)
        table_test.go:25: Expected 6, got 5
FAIL
exit status 1
FAIL    practice/table  0.001s
```

- But if we changed it to:

```go
case "*":
		return num1 * num2, nil
```

```sh
table ➤ go test
PASS
ok      practice/table  0.001s
```

